File: tools/sentinel/src/lib/tools/ebook-refinery.ts
/**
 * 📚 Ebook Refinery — Convert EPUB to atomic Markdown chapters
 * Replaces tools/ebook_refinery/script.py
 *
 * Requires: pandoc installed on system (brew install pandoc)
 * Usage (direct): tsx src/lib/tools/ebook-refinery.ts <file.epub> [output_dir]
 */

import { readFile, writeFile, mkdir, rm } from 'fs/promises';
import { existsSync } from 'fs';
import { join, basename, extname } from 'path';
import { execa } from 'execa';
import { config } from 'dotenv';

config();

function cleanContent(content: string): string {
    // Remove [text]{...} style tags (Calibre artifacts)
    let result = content;
    result = result.replace(/\[([^\]]*)\]\{[^}]*\}/g, '$1');
    result = result.replace(/\[\]\{[^}]*\}/g, '');
    result = result.replace(/\{#[^}]*\}/g, '');
    result = result.replace(/<\/?span[^>]*>/g, '');
    result = result.replace(/<\/?div[^>]*>/g, '');

    // Collapse excessive blank lines
    const lines = result.split('\n');
    const output: string[] = [];
    let lastEmpty = false;
    for (const line of lines) {
        if (line.trim()) {
            output.push(line.trim());
            lastEmpty = false;
        } else if (!lastEmpty) {
            output.push('');
            lastEmpty = true;
        }
    }

    return output.join('\n');
}

function safeFilename(title: string): string {
    return title.replace(/[/:*?"<>|]/g, '_').trim();
}

export async function ebookRefinery(epubPath: string, outputBaseDir?: string): Promise<string> {
    if (!existsSync(epubPath)) throw new Error(`File not found: ${epubPath}`);

    // Check pandoc
    const pandocCheck = await execa('which', ['pandoc'], { reject: false });
    if (pandocCheck.exitCode !== 0) throw new Error('pandoc not found. Install with: brew install pandoc');

    const vaultDir = process.env.GEMINI_WORK_DIR;
    const bookName = basename(epubPath, extname(epubPath));
    const baseDir = outputBaseDir
        ?? (vaultDir ? join(vaultDir, 'resource') : 'resource');
    const outputDir = join(baseDir, `${bookName}全集`);

    await mkdir(outputDir, { recursive: true });

    // 1. Pandoc EPUB → Markdown
    const tempMd = join(outputDir, '_full_temp.md');
    console.log(`[EbookRefinery] Converting EPUB → Markdown...`);
    await execa('pandoc', [epubPath, '-o', tempMd]);

    const fullContent = await readFile(tempMd, 'utf-8');
    await rm(tempMd);

    // 2. Split by H1 headings
    const parts = fullContent.split(/^(# .+)$/m);
    const tocLines: string[] = [`# ${bookName} - 目录\n`];
    let chapterCount = 0;

    // parts = [preamble, h1, body, h1, body, ...]
    for (let i = 1; i < parts.length; i += 2) {
        const titleLine = parts[i].trim();
        const body = parts[i + 1] ?? '';

        const cleanTitle = titleLine
            .replace(/\{#.*?\}/g, '')
            .replace(/^#\s*/, '')
            .trim();
        const safeTitle = safeFilename(cleanTitle);

        const fullText = `# ${cleanTitle}\n\n${body}`;
        const cleaned = cleanContent(fullText);

        chapterCount++;
        const fileName = `${String(chapterCount).padStart(3, '0')}_${safeTitle}.md`;
        const filePath = join(outputDir, fileName);

        await writeFile(filePath, `---\ntitle: ${cleanTitle}\ntype: chapter\nbook: ${bookName}\n---\n\n${cleaned}`, 'utf-8');
        tocLines.push(`- [[${fileName}|${cleanTitle}]]`);
    }

    // 3. Write TOC
    await writeFile(join(outputDir, '00_目录.md'), tocLines.join('\n'), 'utf-8');

    return `📚 炼制完成！共 ${chapterCount} 个章节\n📂 输出目录：${outputDir}`;
}

// Run directly
if (import.meta.url === `file://${process.argv[1]}`) {
    const epubPath = process.argv[2];
    if (!epubPath) { console.error('Usage: tsx ebook-refinery.ts <file.epub> [output_dir]'); process.exit(1); }
    console.log(`📚 Ebook Refinery starting...`);
    const result = await ebookRefinery(epubPath, process.argv[3]);
    console.log(result);
}

File: neo-tools/refinery/ebook_refinery.py
import os
import sys
import subprocess
from pathlib import Path
import re
from dotenv import load_dotenv

def clean_content(content: str) -> str:
    result = re.sub(r'\[([^\]]*)\]\{[^}]*\}', r'\1', content)
    result = re.sub(r'\[\]\{[^}]*\}', '', result)
    result = re.sub(r'\{#[^}]*\}', '', result)
    result = re.sub(r'</?(span|div)[^>]*>', '', result)
    
    lines = result.split('\n')
    output = []
    last_empty = False
    
    for line in lines:
        stripped = line.strip()
        if stripped:
            output.append(stripped)
            last_empty = False
        elif not last_empty:
            output.append('')
            last_empty = True
            
    return '\n'.join(output)

def safe_filename(title: str) -> str:
    return re.sub(r'[/:*?"<>|]', '_', title).strip()

def ebook_refinery(epub_path: str, output_base_dir: str | None = None) -> str:
    load_dotenv()
    
    epub_file = Path(epub_path)
    if not epub_file.exists():
        raise RuntimeError(f"File not found: {epub_path}")
        
    try:
        subprocess.run(['pandoc', '--version'], capture_output=True, check=True)
    except FileNotFoundError:
        raise RuntimeError("pandoc not found. Install with: brew install pandoc")
        
    vault_dir = os.environ.get("GEMINI_WORK_DIR")
    book_name = epub_file.stem
    
    if output_base_dir:
        base_dir = Path(output_base_dir)
    elif vault_dir:
        base_dir = Path(vault_dir) / 'source'
    else:
        base_dir = Path('source')
        
    output_dir = base_dir / f"{book_name}全集"
    output_dir.mkdir(parents=True, exist_ok=True)
    
    temp_md = output_dir / '_full_temp.md'
    print("[EbookRefinery] Converting EPUB -> Markdown...")
    subprocess.run(['pandoc', str(epub_file), '-o', str(temp_md)], check=True)
    
    full_content = temp_md.read_text(encoding='utf-8')
    temp_md.unlink()
    
    parts = re.split(r'^(# .+)$', full_content, flags=re.MULTILINE)
    toc_lines = [f"# {book_name} - 目录\n"]
    chapter_count = 0
    
    # parts looks like: [preamble, title1, body1, title2, body2...]
    for i in range(1, len(parts), 2):
        title_line = parts[i].strip()
        body = parts[i+1] if i + 1 < len(parts) else ''
        
        clean_title = re.sub(r'\{#.*?\}', '', title_line)
        clean_title = re.sub(r'^#\s*', '', clean_title).strip()
        safe_title = safe_filename(clean_title)
        
        full_text = f"# {clean_title}\n\n{body}"
        cleaned = clean_content(full_text)
        
        chapter_count += 1
        file_name = f"{chapter_count:03d}_{safe_title}.md"
        file_path = output_dir / file_name
        
        frontmatter = f"---\ntitle: {clean_title}\ntype: chapter\nbook: {book_name}\n---\n\n"
        file_path.write_text(frontmatter + cleaned, encoding='utf-8')
        
        toc_lines.append(f"- [[{file_name}|{clean_title}]]")
        
    toc_path = output_dir / '目录.md'
    toc_path.write_text('\n'.join(toc_lines), encoding='utf-8')
    
    return f"📚 炼制完成！共 {chapter_count} 个章节\n📂 输出目录：{output_dir}"

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python ebook_refinery.py <file.epub> [output_dir]", file=sys.stderr)
        sys.exit(1)
        
    print("📚 Ebook Refinery starting...")
    try:
        epub_arg = sys.argv[1]
        out_arg = sys.argv[2] if len(sys.argv) > 2 else None
        print(ebook_refinery(epub_arg, out_arg))
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)
